File: android/src/main/java/org/itrabbit/rnwebgl2/RNWebGLContext.java
package org.itrabbit.rnwebgl2;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactContext;

import org.itrabbit.rnwebgl2.utils.FileSystemUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.lang.ref.WeakReference;
import java.nio.IntBuffer;
import java.util.HashMap;
import java.util.Map;

import static android.opengl.GLES30.*;
import static org.itrabbit.rnwebgl2.RNWebGL.*;


public class RNWebGLContext {
    private int mCtxId = -1;

    private final RNWebGLObjectManager mManager;

    public RNWebGLContext(RNWebGLObjectManager manager) {
        super();
        mManager = manager;
    }

    public int getContextId() {
        return mCtxId;
    }

    public void initialize(final Runnable completionCallback) {
        ReactContext reactContext = mManager.getContext();
        long jsContextRef = reactContext.getJavaScriptContextHolder().get();
        mCtxId = RNWebGLContextCreate(jsContextRef);
        mManager.saveContext(this);
        completionCallback.run();
    }

    public void flush() {
        if (mCtxId > 0) {
            RNWebGLContextFlush(mCtxId);
            if (RNWebGLContextNeedsRedraw(mCtxId)) {
                RNWebGLContextDrawEnded(mCtxId);
            }
        }
    }


    public void destroy() {
        mManager.deleteContextWithId(mCtxId);
        mManager.unloadWithCtxId(mCtxId);
        RNWebGLContextDestroy(mCtxId);
    }

    // must be called in GL thread
    public Map<String, Object> getViewportRect() {
        int[] viewport = new int[4];
        glGetIntegerv(GL_VIEWPORT, viewport, 0);

        Map<String, Object> results = new HashMap<>();
        results.put("x", viewport[0]);
        results.put("y", viewport[1]);
        results.put("width", viewport[2]);
        results.put("height", viewport[3]);

        return results;
    }

    public void takeSnapshot(final Map<String, Object> options, final Context context, final Promise promise) {
        flush();
        RNWebGLView.runOnGLThread(mCtxId, new Runnable() {
            @SuppressWarnings({"unchecked", "ConstantConditions"})
            @Override
            public void run() {
                Map<String, Object> rect = options.containsKey("rect") ? (Map<String, Object>) options.get("rect") : getViewportRect();
                Boolean flip = options.containsKey("flip") && (Boolean) options.get("flip");
                String format = options.containsKey("format") ? (String) options.get("format") : null;
                int compressionQuality = options.containsKey("compress") ? (int) (100.0 * (Double) options.get("compress")) : 100;

                int x = castNumberToInt(rect.get("x"));
                int y = castNumberToInt(rect.get("y"));
                int width = castNumberToInt(rect.get("width"));
                int height = castNumberToInt(rect.get("height"));

                // Save surrounding framebuffer
                int[] prevFramebuffer = new int[1];
                glGetIntegerv(GL_FRAMEBUFFER_BINDING, prevFramebuffer, 0);

                // Set source framebuffer that we take snapshot from
                int sourceFramebuffer = prevFramebuffer[0];
                Map<String, Object> framebufferMap = options.containsKey("framebuffer") ? (Map<String, Object>) options.get("framebuffer") : null;

                if (framebufferMap != null && framebufferMap.containsKey("id")) {
                    Integer framebufferId = castNumberToInt(framebufferMap.get("id"));
                    sourceFramebuffer = RNWebGLContextGetObject(mCtxId, framebufferId);
                }

                // Bind source framebuffer
                glBindFramebuffer(GL_FRAMEBUFFER, sourceFramebuffer);

                // Allocate pixel buffer and read pixels
                final int[] dataArray = new int[width * height];
                final IntBuffer dataBuffer = IntBuffer.wrap(dataArray);
                dataBuffer.position(0);
                glReadPixels(x, y, width, height, GL_RGBA, GL_UNSIGNED_BYTE, dataBuffer);

                // Restore surrounding framebuffer
                glBindFramebuffer(GL_FRAMEBUFFER, prevFramebuffer[0]);

                new TakeSnapshot(context, width, height, flip, format, compressionQuality, dataArray, promise)
                        .executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
            }
        });
    }

    private static class TakeSnapshot extends AsyncTask<Void, Void, Void> {
        private final WeakReference<Context> mContext;
        private final int mWidth;
        private final int mHeight;
        private final boolean mFlip;
        private final String mFormat;
        private final int mCompress;
        private final int[] mDataArray;
        private final Promise mPromise;

        TakeSnapshot(Context context, int width, int height, boolean flip, String format, int compress, int[] dataArray, Promise promise) {
            mContext = new WeakReference<>(context);
            mWidth = width;
            mHeight = height;
            mFlip = flip;
            mFormat = format;
            mCompress = compress;
            mDataArray = dataArray;
            mPromise = promise;
        }

        @Override
        protected Void doInBackground(Void... params) {
            // Convert RGBA data format to bitmap's ARGB
            for (int i = 0; i < mHeight; i++) {
                for (int j = 0; j < mWidth; j++) {
                    int offset = i * mWidth + j;
                    int pixel = mDataArray[offset];
                    int blue = (pixel >> 16) & 0xff;
                    int red = (pixel << 16) & 0x00ff0000;
                    mDataArray[offset] = (pixel & 0xff00ff00) | red | blue;
                }
            }

            // Create Bitmap and flip
            Bitmap bitmap = Bitmap.createBitmap(mDataArray, mWidth, mHeight, Bitmap.Config.ARGB_8888);

            if (!mFlip) {
                // the bitmap is automatically flipped on Android, however we may want to unflip it
                // in case we take a snapshot from framebuffer that is already flipped
                Matrix flipMatrix = new Matrix();
                flipMatrix.postScale(1, -1, mWidth / 2, mHeight / 2);
                bitmap = Bitmap.createBitmap(bitmap, 0, 0, mWidth, mHeight, flipMatrix, true);
            }

            // Write bitmap to file
            String path = null;
            String extension = ".jpeg";
            FileOutputStream output = null;
            Bitmap.CompressFormat compressFormat = Bitmap.CompressFormat.JPEG;

            if (mFormat != null && mFormat.equals("png")) {
                compressFormat = Bitmap.CompressFormat.PNG;
                extension = ".png";
            }

            Context context = mContext.get();

            if (context == null) {
                mPromise.reject("E_GL_CONTEXT_DESTROYED", "Context has been garbage collected.");
                return null;
            }

            try {
                path = FileSystemUtils.generateOutputPath(context.getCacheDir(), "GLView", extension);
                output = new FileOutputStream(path);
                bitmap.compress(compressFormat, mCompress, output);
                output.flush();
                output.close();
                output = null;

            } catch (Exception e) {
                e.printStackTrace();
                mPromise.reject("E_GL_CANT_SAVE_SNAPSHOT", e.getMessage());
            }

            if (output == null) {
                // Return result object which imitates Expo.Asset so it can be used again to fill the texture
                Bundle result = new Bundle();
                String fileUri = Uri.fromFile(new File(path)).toString();

                result.putString("uri", fileUri);
                result.putString("localUri", fileUri);
                result.putInt("width", mWidth);
                result.putInt("height", mHeight);

                mPromise.resolve(result);
            }
            return null;
        }
    }

    // Solves number casting problem as number values can come as Integer or Double.
    private int castNumberToInt(Object value) {
        if (value instanceof Double) {
            return ((Double) value).intValue();
        }
        return (Integer) value;
    }
}

File: ios/RNWebGLContext.m
#import "RNWebGLContext.h"

#include <OpenGLES/ES3/gl.h>
#include <OpenGLES/ES3/glext.h>

#define BLOCK_SAFE_RUN(block, ...) block ? block(__VA_ARGS__) : nil

@interface RNWebGLContext ()

@property (nonatomic, strong) dispatch_queue_t glQueue;
@property (nonatomic, weak) RNWebGLObjectManager *objectManager;

@end

@interface RCTBridge ()

- (JSGlobalContextRef)jsContextRef;
- (void)dispatchBlock:(dispatch_block_t)block queue:(dispatch_queue_t)queue;

@end

@implementation RNWebGLContext

- (instancetype)initWithDelegate:(id<RNWebGLContextDelegate>)delegate andObjectMananger:(nonnull RNWebGLObjectManager *)objectManager {
    if (self = [super init]) {
        self.delegate = delegate;
        _objectManager = objectManager;
        _glQueue = dispatch_queue_create("host.rn.gl", DISPATCH_QUEUE_SERIAL);
        _eaglCtx = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES3] ?: [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2];
    }
    return self;
}

- (BOOL)isInitialized {
    return _contextId != 0;
}

- (EAGLContext *)createSharedEAGLContext {
    return [[EAGLContext alloc] initWithAPI:[_eaglCtx API] sharegroup:[_eaglCtx sharegroup]];
}

- (void)runInEAGLContext:(EAGLContext*)context callback:(void(^)(void))callback
{
    [EAGLContext setCurrentContext:context];
    callback();
    glFlush();
    [EAGLContext setCurrentContext:nil];
}

- (void)runAsync:(void(^)(void))callback {
    if (_glQueue) {
        dispatch_async(_glQueue, ^{
            [self runInEAGLContext:self->_eaglCtx callback:callback];
        });
    }
}

- (void)initialize:(void(^)(BOOL))callback {
    RCTBridge *bridge = _objectManager.bridge;
    if (!bridge.executorClass || [NSStringFromClass(bridge.executorClass) isEqualToString:@"RCTJSCExecutor"]) {
        __weak __typeof__(self) weakSelf = self;
        __weak __typeof__(bridge) weakBridge = bridge;
        [bridge dispatchBlock:^{
            __typeof__(self) self = weakSelf;
            __typeof__(bridge) bridge = weakBridge;
            if (!self || !bridge || !bridge.valid) {
                BLOCK_SAFE_RUN(callback, NO);
                return;
            }
            JSGlobalContextRef jsContextRef = [bridge jsContextRef];
            if (!jsContextRef) {
                BLOCK_SAFE_RUN(callback, NO);
                RCTLogError(@"RNWebGL: The React Native bridge unexpectedly does not have a JavaScriptCore context.");
                return;
            }
            self->_contextId = RNWebGLContextCreate(jsContextRef);
            [self->_objectManager saveContext:self];
            /*
            RNWebGLContextSetFlushMethodObjc(self->_contextId, ^{
                [self flush];
            });
            */
            
            if ([self.delegate respondsToSelector:@selector(glContextInitialized:)]) {
                [self.delegate glContextInitialized:self];
            }
            BLOCK_SAFE_RUN(callback, YES);
        } queue:RCTJSThread];
        return;
    }
    BLOCK_SAFE_RUN(callback, NO);
    RCTLogError(@"RNWebGL: Can only run on JavaScriptCore! Do you have 'Remote Debugging' enabled in your app's Developer Menu (https://facebook.github.io/react-native/docs/debugging.html)? RNWebGL is not supported while using Remote Debugging, you will need to disable it to use RNWebGL.");
}

- (void)flush {
    [self runAsync:^{
        RNWebGLContextFlush(self->_contextId);
        if ([self.delegate respondsToSelector:@selector(glContextFlushed:)]) {
            [self.delegate glContextFlushed:self];
        }
    }];
}

- (void)destroy {
    [self runAsync:^{
        if ([self.delegate respondsToSelector:@selector(glContextWillDestroy:)]) {
            [self.delegate glContextWillDestroy:self];
        }
        // Flush all the stuff
        RNWebGLContextFlush(self->_contextId);
        // Destroy JS binding
        RNWebGLContextDestroy(self->_contextId);
        // Unload objects
        [self->_objectManager unloadWithCtxId:self->_contextId];
        // Remove from dictionary of contexts
        [self->_objectManager deleteContextWithId:@(self->_contextId)];
    }];
}

# pragma mark - snapshots

// Saves the contents of the framebuffer to a file.
// Possible options:
// - `flip`: if true, the image will be flipped vertically.
// - `framebuffer`: WebGLFramebuffer that we will be reading from. If not specified, the default framebuffer for this context will be used.
// - `rect`: { x, y, width, height } object used to crop the snapshot.
// - `format`: "jpeg" or "png" - specifies what type of compression and file extension should be used.
// - `compress`: A value in 0 - 1 range specyfing compression level. JPEG format only.
- (void)takeSnapshotWithOptions:(nonnull NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject {
    [self flush];
    [self runAsync:^{
        NSDictionary *rect = options[@"rect"] ?: [self currentViewport];
        BOOL flip = options[@"flip"] != nil && [options[@"flip"] boolValue];
        NSString *format = options[@"format"];
        
        int x = [rect[@"x"] intValue];
        int y = [rect[@"y"] intValue];
        int width = [rect[@"width"] intValue];
        int height = [rect[@"height"] intValue];
        
        // Save surrounding framebuffer
        GLint prevFramebuffer;
        glGetIntegerv(GL_FRAMEBUFFER_BINDING, &prevFramebuffer);
        
        // Set source framebuffer that we take snapshot from
        GLint sourceFramebuffer = 0;
        
        if (options[@"framebuffer"] && options[@"framebuffer"][@"id"]) {
            int RNWebGLFramebufferId = [options[@"framebuffer"][@"id"] intValue];
            sourceFramebuffer = RNWebGLContextGetObject(self.contextId, RNWebGLFramebufferId);
        } else {
            // headless context doesn't have default framebuffer, so we use the current one
            sourceFramebuffer = [self defaultFramebuffer] || prevFramebuffer;
        }
        if (sourceFramebuffer == 0) {
            reject(
                   @"RNWEBGL_NO_FRAMEBUFFER",
                   nil,
                   RCTErrorWithMessage(@"No framebuffer bound. Create and bind one to take a snapshot from it.")
                   );
            return;
        }
        if (width <= 0 || height <= 0) {
            reject(
                   @"RNWEBGL_INVALID_VIEWPORT",
                   nil,
                   RCTErrorWithMessage(@"Rect's width and height must be greater than 0. If you didn't set `rect` option, check if the viewport is set correctly.")
                   );
            return;
        }
        
        // Bind source framebuffer
        glBindFramebuffer(GL_FRAMEBUFFER, sourceFramebuffer);
        
        // Allocate pixel buffer and read pixels
        NSInteger dataLength = width * height * 4;
        GLubyte *buffer = (GLubyte *) malloc(dataLength * sizeof(GLubyte));
        glReadBuffer(GL_COLOR_ATTACHMENT0);
        glReadPixels(x, y, width, height, GL_RGBA, GL_UNSIGNED_BYTE, buffer);
        
        // Create CGImage
        CGDataProviderRef providerRef = CGDataProviderCreateWithData(NULL, buffer, dataLength, NULL);
        CGColorSpaceRef colorspaceRef = CGColorSpaceCreateDeviceRGB();
        CGImageRef imageRef = CGImageCreate(width, height, 8, 32, width * 4, colorspaceRef, kCGBitmapByteOrder32Big | kCGImageAlphaPremultipliedLast,
                                            providerRef, NULL, true, kCGRenderingIntentDefault);
        
        // Begin image context
        CGFloat scale = [RNWebGLContext screenScale];
        NSInteger widthInPoints = width / scale;
        NSInteger heightInPoints = height / scale;
        UIGraphicsBeginImageContextWithOptions(CGSizeMake(widthInPoints, heightInPoints), NO, scale);
        
        // Flip and draw image to CGImage
        CGContextRef cgContext = UIGraphicsGetCurrentContext();
        if (flip) {
            CGAffineTransform flipVertical = CGAffineTransformMake(1, 0, 0, -1, 0, heightInPoints);
            CGContextConcatCTM(cgContext, flipVertical);
        }
        CGContextDrawImage(cgContext, CGRectMake(0.0, 0.0, widthInPoints, heightInPoints), imageRef);
        
        // Retrieve the UIImage from the current context
        UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        
        // Cleanup
        free(buffer);
        CFRelease(providerRef);
        CFRelease(colorspaceRef);
        CGImageRelease(imageRef);
        
        // Write image to file
        NSData *imageData;
        NSString *extension;
        
        if ([format isEqualToString:@"png"]) {
            imageData = UIImagePNGRepresentation(image);
            extension = @".png";
        } else {
            float compress = 1.0;
            if (options[@"compress"] != nil) {
                compress = [(NSString *)options[@"compress"] floatValue];
            }
            imageData = UIImageJPEGRepresentation(image, compress);
            extension = @".jpeg";
        }
        
        NSString *filePath = [self generateSnapshotPathWithExtension:extension];
        [imageData writeToFile:filePath atomically:YES];
        
        // Restore surrounding framebuffer
        glBindFramebuffer(GL_FRAMEBUFFER, prevFramebuffer);
        
        // Return result object which imitates Expo.Asset so it can be used again to fill the texture
        NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
        NSString *fileUrl = [[NSURL fileURLWithPath:filePath] absoluteString];
        
        result[@"uri"] = fileUrl;
        result[@"localUri"] = fileUrl;
        result[@"width"] = @(width);
        result[@"height"] = @(height);
        
        resolve(result);
    }];
}

- (NSDictionary *)currentViewport {
    GLint viewport[4];
    glGetIntegerv(GL_VIEWPORT, viewport);
    return @{ @"x": @(viewport[0]), @"y": @(viewport[1]), @"width": @(viewport[2]), @"height": @(viewport[3]) };
}

- (GLint)defaultFramebuffer {
    if ([self.delegate respondsToSelector:@selector(glContextGetDefaultFramebuffer)]) {
        return [self.delegate glContextGetDefaultFramebuffer];
    }
    return 0;
}

- (NSString *)getPathForDirectory:(int)directory {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(directory, NSUserDomainMask, YES);
    return [paths firstObject];
}

- (NSString *)generateSnapshotPathWithExtension:(NSString *)extension {
    NSString *directory = [[self getPathForDirectory: NSCachesDirectory] stringByAppendingPathComponent:@"GLView"];
    NSString *fileName = [[[NSUUID UUID] UUIDString] stringByAppendingString:extension];
    
    [[NSFileManager defaultManager] createDirectoryAtPath:directory withIntermediateDirectories:YES attributes:@{} error:nil];
    
    return [directory stringByAppendingPathComponent:fileName];
}

#pragma mark - Copy from RN

+ (BOOL)isMainQueue {
    static void *mainQueueKey = &mainQueueKey;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        dispatch_queue_set_specific(dispatch_get_main_queue(),
                                    mainQueueKey, mainQueueKey, NULL);
    });
    return dispatch_get_specific(mainQueueKey) == mainQueueKey;
}

+ (void)unsafeExecuteOnMainQueueOnceSync:(dispatch_once_t *)onceToken block:(dispatch_block_t)block {
    // The solution was borrowed from a post by Ben Alpert:
    // https://benalpert.com/2014/04/02/dispatch-once-initialization-on-the-main-thread.html
    // See also: https://www.mikeash.com/pyblog/friday-qa-2014-06-06-secrets-of-dispatch_once.html
    if ([self isMainQueue]) {
        dispatch_once(onceToken, block);
    } else {
        if (DISPATCH_EXPECT(*onceToken == 0L, NO)) {
            dispatch_sync(dispatch_get_main_queue(), ^{
                dispatch_once(onceToken, block);
            });
        }
    }
}

+ (CGFloat)screenScale {
    static dispatch_once_t onceToken;
    static CGFloat scale;
    [self unsafeExecuteOnMainQueueOnceSync:&onceToken block:^{
        scale = [UIScreen mainScreen].scale;
    }];
    return scale;
}

@end
